### ISSUE TITLE
custom vars list

### ISSUE THREAD

**Issue by cbnorman** [Icinga/icingaweb2-module-director#975]
I would like to be able to list all the puppet groups applied to a host. this would be handy to see in the web page but even more beneficial so that we know when a host is configured to do something via a puppet group the associated monitoring would also be in effect.

## Expected Behavior
here is what i would like to achieve with director:
```
object Host "test-host"{
  vars.address = "8.8.8.8"
  check_command = "ping4"
  vars.puppet_groups = ["REBOOT_ENABLED", "BASE_MANAGED"]
  }
```

![image](https://user-images.githubusercontent.com/10809716/27384785-20d31778-5688-11e7-93f4-5b0c4aaa0b48.png)

## what I've tried
I've tried creating a list in director for one host and used that in the host sync rule by adding a modifier. 
![image](https://user-images.githubusercontent.com/10809716/27384876-80b97538-5688-11e7-8abd-8290711fcc6b.png)

![image](https://user-images.githubusercontent.com/10809716/27384902-98f56184-5688-11e7-85b3-a1abb8ec2195.png)

here is when running the sync rule:
![image](https://user-images.githubusercontent.com/10809716/27384972-dc0daaee-5688-11e7-81e9-b326f44ddd34.png)

## Your Environment

* Director version (System - About):
Version	1.3.1
Git commit	4d240976a3e1ffe1eb9555fa23067da4aed25a58
* Icinga Web 2 version and modules (System - About):
2.4.0
Git commit
f7a8cf8a6ee8d94ba2997348907fbbc9521aa687
* Icinga 2 version (`icinga2 --version`):
icinga2 --version
icinga2 - The Icinga 2 network monitoring daemon (version: v2.6.3-151-g46900ce)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: CentOS Linux
  Platform version: 7 (Core)
  Kernel: Linux
  Kernel version: 3.10.0-327.22.2.el7.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown

* Operating System and version:
hostnamectl
   Static hostname: elkmon132.suntradingllc.com
         Icon name: computer-vm
           Chassis: vm
        Machine ID: 53cc3ca755e749bf94f2e4e98374ea66
           Boot ID: 9c4e2673006f409c9aa37e5878e784b0
    Virtualization: kvm
  Operating System: CentOS Linux 7 (Core)
       CPE OS Name: cpe:/o:centos:centos:7
            Kernel: Linux 3.10.0-327.22.2.el7.x86_64
      Architecture: x86-64

* Webserver, PHP versions:

**Comment by cbnorman** [comment]
im thinking using datalist value as array type is not currently supported. is there another way to get these information as a list assigned to hosts using director?

**Comment by cbnorman** [comment]
https://github.com/Icinga/icinga2/issues/1717 shows the feature i want to use more clearly, maybe its just not possible with director yet?

**Comment by capull0** [comment]
i achieved this, creating the object over the Director API as json, or just update the vars.puppet_groups over the API. You can't modify this variable over Icingaweb2 director then, only via the Director API.

https://github.com/Icinga/icingaweb2-module-director/blob/master/doc/70-REST-API.md

`POST director/host?name=test-host`

`{"vars.puppet_groups":[\"REBOOT_ENABLED\", \"BASE_MANAGED\"]}`

don't know if this helps.

**Comment by cbnorman** [comment]
thanks @capull0 im aware of the director api and prior to director used the icinga api for all object creation. director db sync keeps it clear for everyone (none of my hacky work) and i can dont have to maintain extra sync scheduling. @Thomas-Gelf do you know if this feature will be released soon or am i better off going down the api route?

thanks all

**Comment by Thomas-Gelf** [comment]
@cbnorman: please help me to better understand your issue. Or let me explain what I understood so far:
* you want to use a Data List based Data Field to fill an Array property. The fix to this should be allowing you to set a "Target Data Type" when defining such a Field. You can have a look at the `DirectorObject` Data Type - it would work like that.
* but what I do not get is your problem with Import/Sync, at for what I understood so far that's what you're using to fill your vars. And Sync is perfectly able to create Arrays and even Dictionaries (Hashes) - a feature not even available in the GUI. Just, your Import Source must be able to ship structured data. CSV files or SQL databases are for example not able to do so. When such sources are involved, I'm usually doing some GROUP_CONCAT magic and then transform the result with the Split modifier.

And honestly I absolutely did not understand what you wanted to achieve with the Map lookup and your list. Could you please give me more details on your scenario to give me a better understanding of your big picture?

Cheers,
Thomas

**Comment by cbnorman** [comment]
HI @Thomas-Gelf 

sorry as i have got more desperate and posted more and more its got quite unclear as to what I'm trying to do, maybe i don't need to use datalists in this case as you showed me in #931 which I'm using quite a bit now so thanks for that.

Basically I have a database that has a number of rows per hosts each one showing its puppet group membership so like:
```select host, group from puppet_groups;
host               group
hosta             groupname1
hosta             groupname5
hosta             groupname6
hosta             groupname7
hostb             groupname1
hostb             groupname2
hostc             groupname9
hostc             groupname10
hostc             groupname11```

First of all I cant import it like this as host isn't a unique field, I was thinking of resolving this by creating a modifier making another field based on the host and group or maybe changing the sql query. From this data ultimately id like to add the group membership as an array/dictionary to a custom var for each host. i was thinking i would need to create a datalist and then map this somehow as a array in the sync rule. i have seen a couple of other tickets around supporting arrays in data lists which you seem to be working on like #33

so it sounds like i should change the sql query to group them and then split the result in a modifier. is there any documentation on this yet? im going to give this a try tomorrow and ill post my results if i hopefully get it working. 

Thanks for the pointers, Really Appreciated.

**Comment by Thomas-Gelf** [comment]
@cbnorman: did it work for you? For the last example just GROUP_CONCAT() the groups in your SQL query and split them into an array with a property modifier.

**Comment by cbnorman** [comment]
Hi Tom. Yes that worked, although the db I pulled it from is one which I didn't want all the hosts so there was a load that didn't have a check command etc because that's configured in another source. I've since manually set the new hosts that it want to create to disabled = yes. So there in director just not being processed on deploy. Occasionally I have to disable a new one here and there, would be nice if we could pull from another source and have it merge with existing hosts only. 

I've got what I need though so this can be closed if setting disabled=yes is the best way to deal with the extra hosts?

Thanks agian. 

Chris
